### ISSUE TITLE
Generalize support for exit events

### ISSUE THREAD

**Issue by powerjg** [gem5/gem5#474]
Note: This issue is a brain dump and will likely be updated.

## Problem

Today, the exit events in gem5 are not very flexible. Most of the information that moves from the C++ event queue to Python is implemented via strings (e.g., "user interrupt received"). Some other places for improvements

- `SIGUSR1` is hardcoded to dump stats
- `SIGUSR2` is hardcoded to dump and reset stats
- `SIGINT` does not always actually exit the simulation

The other main problem we are trying to solve is enabling users to extend the exit event types.
For instance, exiting for SimPoints, LoopPoints, or checking on the current state of the simulation.

## Some steps

- [ ] Fix SIGINT to work when executing on the non-main event queue
  - #476 
- [ ] Refactor `SIGUSR1/2` to exit to python and then execute the dump/reset stats in Python
- [ ] Create a different class for user-defined exit events
- [ ] Also allow for return python-implemented exit events
- [ ] Implement a new signal (maybe using `SIGRT`?) so that Python can send a signal to stop the simulation

**Comment by andysan** [comment]
Looking at the code, it's pretty clear that there are race conditions in the way asynchronous events are notified. I think we might need to rewrite that code to use a flag field based on an std::atomic_uint32_t to ensure we can atomically read and clear multiple flags.

**Comment by andysan** [comment]
I have been thinking a bit about this and replacing the strings with proper objects would be a big improvement. Ideally, I would like to see something looking a bit like exceptions on the Python side. We currently have two big problems. It's very easy to lose exit events by incorrectly ignoring the return value of `m5.simulate()`. The other problem is drain where we need to run the simulator, but we actually ignore exit events. Treating them as exceptions would be cleaner.

I think there is a path that maintains backwards compatibility here. We could implement `__str__` in the exit event wrapper to maintain backwards compatibility. Exit event handling could be handled internally by a global callback (that can be overridden as a keyword argument to `simulate`?)  that either returns the event, returns None (ignore the event), or throws the event as an exception. We could provide two implementation, one that always throws the event and one that always returns the event. The default would be to return the event to maintain backwards compatibility.

We should probably avoid using a signal to stop the simulator from Python, at least SIGRT which is used by KVM. A better option would probably be to use a socket (which could use a SIGIO in gem5) which would give us a bit more flexibility to control the simulator from the outside.

**Comment by trollet91** [comment]
> The other problem is drain where we need to run the simulator, but we actually ignore exit events

Is there a reason for not re-injecting the events that where caught during the drain?
Ex: user set a checkpoint event at tick 1000 and exit at tick 1001.
In my perspective, it would make sense to execute the exit event once the drain + checkpoint is done (This is what we do in our gem5 setup). Would it break other use-cases?

Related question: is there a use-case of the simulator where the eventq really needs to be in LIFO for performance or correctness?  A FIFO would make the behaviour more natural for the user (ex: if user set a checkpoint then exit on the same tick, he expect the checkpoint to happen before.)

**Comment by andysan** [comment]
> > The other problem is drain where we need to run the simulator, but we actually ignore exit events
> 
> Is there a reason for not re-injecting the events that where caught during the drain? Ex: user set a checkpoint event at tick 1000 and exit at tick 1001. In my perspective, it would make sense to execute the exit event once the drain + checkpoint is done (This is what we do in our gem5 setup). Would it break other use-cases?

This is functionally similar to what I did in the past, but instead of re-injecting them I just returned a list of events from `drain`. In most cases, re-injecting events should be fine. However, there are cases where it can lead to undesirable outcomes. These are mostly related to instruction count triggered exits and m5op triggered exits combined with setups where draining takes thousands of instructions. If the exit is delayed, you probably need to know about it or you'll get weird systematic errors (e.g., sampling errors). This gets particularly weird if you have two instruction count exits close together and they just happened to collide because they were both delayed by a drain.

Treating exits as exceptions in drain would neatly solve this problem, but it also means that you'd return to the simulation script in a partially drained state and it wouldn't be backwards compatible. A good solution would maybe be to have two modes when calling drain. One that is backwards compatible and re-injects exit events (we could just add a pending queue and do this entirely in Python) and one that throws exceptions.

> Related question: is there a use-case of the simulator where the eventq really needs to be in LIFO for performance or correctness? A FIFO would make the behaviour more natural for the user (ex: if user set a checkpoint then exit on the same tick, he expect the checkpoint to happen before.)

I don't think the API guarantees any particular order for events of the same priority at the same tick. There might be an order as an implementation artefact, but I wouldn't rely on it.

**Comment by trollet91** [comment]
>  A good solution would maybe be to have two modes when calling drain. One that is backwards compatible and re-injects exit events (we could just add a pending queue and do this entirely in Python) and one that throws exceptions.

Thanks for the explanation. Seems like a good tradeoff.

>  There might be an order as an implementation artefact, but I wouldn't rely on it.

The problem is that it's not an easy information to convey (other than discovering it while debugging why an event is not executed). So I my case I'm making sure that:
- all user defined events based on ticks must be on different ticks
- all user events based on instructions happens in different cpu cycles (distance between events must be greater than commit width in O3CPU) 

I wonder what would be the downside caused by adding FIFO in the API guarantees.

**Comment by erin-le** [comment]
With the merging of https://github.com/gem5/gem5/pull/1982, these bullet points have either been implemented, or other changes have made them no longer relevant. Closing as complete.
